components/brave_wallet/browser/json_rpc_service.cc

```diff
@@ -1395,6 +1396,190 @@ void JsonRpcService::ContinueGetERC721TokenBalance(
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  // Parse response JSON that wraps the result
+  std::string result;
+  if (!ParseSingleStringResult(body, &result)) {
+    mojom::ProviderError error;
+    std::string error_message;
+    ParseErrorResult<mojom::ProviderError>(body, &error, &error_message);
+    std::move(callback).Run("", error, error_message);
+    return;
+  }
+
+  // Parse the URI
+  GURL url;
+  if (!eth::ParseERC721TokenUri(
+          result, &url)) {  // Either invalid RLP encoding, or invalid URL
+    std::move(callback).Run("", mojom::ProviderError::kParsingError,
+                            l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    return;
+  }
+
+  // Obtain JSON from the URL depending on the scheme.
+  // IPFS, HTTPS, and data URIs are supported.
+  // IPFS and HTTPS URIs require an additional request to fetch the metadata.
+  std::string metadata_json;
+  std::string scheme = url.scheme();
+  if (scheme != url::kDataScheme && scheme != url::kHttpsScheme &&
+      scheme != ipfs::kIPFSScheme) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  } else if (scheme == url::kDataScheme) {
+    if (!ParseDataURIAndExtractJSON(url, &metadata_json)) {
+      std::move(callback).Run(
+          "", mojom::ProviderError::kParsingError,
+          l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+      return;
+    }
+
+    // Sanitize JSON
+    data_decoder::JsonSanitizer::Sanitize(
+        std::move(metadata_json),
+        base::BindOnce(&JsonRpcService::OnSanitzeERC721InLineMetadata,
+                       weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+    return;
+  } else if (scheme == ipfs::kIPFSScheme &&
+             !ipfs::ToConfiguredGatewayURL(&url, prefs_)) {
+    std::move(callback).Run("", mojom::ProviderError::kParsingError,
+                            l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    return;
+  }
+
+  auto internal_callback =
+      base::BindOnce(&JsonRpcService::OnGetERC721MetadataPayload,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+  api_request_helper_->Request("GET", url, "", "", true,
+                               std::move(internal_callback));
+}
+
+void JsonRpcService::OnSanitzeERC721InLineMetadata(
+    GetERC721MetadataCallback callback,
+    data_decoder::JsonSanitizer::Result result) {
+  if (result.error) {
+    VLOG(1) << "Data URI JSON validation error:" << *result.error;
+    std::move(callback).Run("", mojom::ProviderError::kParsingError,
+                            l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    return;
+  }
+
+  std::string metadata_json;
+  if (result.value.has_value()) {
+    metadata_json = result.value.value();
+  }
```

Comment:
We might want to treat empty metadata as a failure case here?

Reply:
I agree.  I think this is the current behavior though since if it is an empty string, the sanitization step will return an error, so we will return kParsingError in the block above.

The reason why we have to do an extra check in OnGetERC721MetadataPayload is because if the api_request_helper encounters any invalid JSON, it doesn't pass back an error, it just passes through an empty string https://github.com/brave/brave-core/blob/0f6d04a10996e0831403780a7109c643d8f168cb/components/api_request_helper/api_request_helper.cc#L26-L28.

I'll add a testcase to demonstrate.